### PR TITLE
AWS: Support S3 DSSE-KMS encryption

### PR DESCRIPTION
**Pull request by aajisaka**
This PR is to support S3 Dual-layer Server Side Encryption (DSSE) for Iceberg tables.

https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingDSSEncryption.html

### CODE REVIEW

File: aws/src/test/java/org/apache/iceberg/aws/s3/signer/TestS3SignRequestParser.java [hunk 1]
```diff
@@ -126,7 +126,7 @@ public void roundTripSerde() {
                     "Content-Type",
                     Arrays.asList("application/json"),
                     "User-Agent",
-                    Arrays.asList("aws-sdk-java/2.20.18", "Linux/5.4.0-126")))
+                    Arrays.asList("aws-sdk-java/2.20.131", "Linux/5.4.0-126")))
```
Comment (nastra) [inline aws/src/test/java/org/apache/iceberg/aws/s3/signer/TestS3SignRequestParser.java:129]: we don't really need to update those. It doesn't matter what headers the test is using

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIOProperties.java [hunk 2]
```diff
@@ -67,6 +67,14 @@ public class S3FileIOProperties implements Serializable {
    */
   public static final String SSE_TYPE_KMS = "kms";
 
+  /**
+   * S3 DSSE-KMS encryption.
+   *
+   * <p>For more details:
+   * https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingDSSEncryption.html
+   */
+  public static final String DSSE_TYPE_KMS = "kms:dsse";
```
Comment (nastra) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIOProperties.java:76]: I think this should be named `kms.dsse`

**Comment by nastra** [comment]
I think it would be better to split this into 2 PRs (one that only updates the AWS SDK version)

File: docs/aws.md [hunk 3]
```diff
@@ -322,13 +322,14 @@ Here are the configurations that users can tune related to this feature:
...
 * [SSE-KMS](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingKMSEncryption.html): Server-Side Encryption with Customer Master Keys (CMKs) Stored in AWS Key Management Service (SSE-KMS) is similar to SSE-S3, but with some additional benefits and charges for using this service. There are separate permissions for the use of a CMK that provides added protection against unauthorized access of your objects in Amazon S3. SSE-KMS also provides you with an audit trail that shows when your CMK was used and by whom. Additionally, you can create and manage customer managed CMKs or use AWS managed CMKs that are unique to you, your service, and your Region.
+* [DSSE-KMS](https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingDSSEncryption.html): Dual-layer Server-Side Encryption with AWS Key Management Service keys (DSSE-KMS) is similar to SSE-KMS, but applies two layers of encryption to objects when they are uploaded to Amazon S3. DSSE-KMS can be used to fulfill compliance standaards that require you to apply multilayer encryption to your data and have full control of your encryption keys.
```
Comment (singhpk234) [inline docs/aws.md:325]: ``
```suggestion
* [DSSE-KMS](https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingDSSEncryption.html): Dual-layer Server-Side Encryption with AWS Key Management Service keys (DSSE-KMS) is similar to SSE-KMS, but applies two layers of encryption to objects when they are uploaded to Amazon S3. DSSE-KMS can be used to fulfill compliance standards that require you to apply multilayer encryption to your data and have full control of your encryption keys.
```

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIOProperties.java [hunk 4]
```diff
@@ -84,9 +92,9 @@ public class S3FileIOProperties implements Serializable {
   public static final String SSE_TYPE_CUSTOM = "custom";
 
   /**
-   * If S3 encryption type is SSE-KMS, input is a KMS Key ID or ARN. In case this property is not
-   * set, default key "aws/s3" is used. If encryption type is SSE-C, input is a custom base-64
-   * AES256 symmetric key.
+   * If S3 encryption type is SSE-KMS or DSSE-KMS, input is a KMS Key ID or ARN. In case this
```
Comment (singhpk234) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIOProperties.java:95]: can rephrase this

**Comment by aajisaka** [comment]
Thank you @nastra and @singhpk234 for your review.

Let me create a separate PR to upgrade AWS SDK version.

**Comment by aajisaka** [comment]
Opened #8379 to upgrade AWS SDK version

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIOProperties.java (hunk 2 above)
Comment (aajisaka) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIOProperties.java]: Thanks. Fixed

File: aws/src/test/java/org/apache/iceberg/aws/s3/signer/TestS3SignRequestParser.java (hunk 1 above)
Comment (aajisaka) [reply aws/src/test/java/org/apache/iceberg/aws/s3/signer/TestS3SignRequestParser.java]: Thanks. Reverted the change

File: docs/aws.md (hunk 3 above)
Comment (aajisaka) [reply docs/aws.md]: Fixed

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIOProperties.java (hunk 4 above)
Comment (aajisaka) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIOProperties.java:95]: What does that exactly mean?

File: python/dev/Dockerfile [hunk 5]
```diff
@@ -39,7 +39,7 @@ WORKDIR ${SPARK_HOME}
 ENV SPARK_VERSION=3.4.1
 ENV ICEBERG_SPARK_RUNTIME_VERSION=3.4_2.12
 ENV ICEBERG_VERSION=1.3.1
-ENV AWS_SDK_VERSION=2.20.18
+ENV AWS_SDK_VERSION=2.20.131
```
Comment (nastra) [inline python/dev/Dockerfile:42]: needs to be reverted

File: aws/src/test/java/org/apache/iceberg/aws/s3/signer/TestS3SignResponseParser.java [hunk 6]
```diff
@@ -74,7 +74,7 @@ public void roundTripSerde() {
                     "Content-Type",
                     Arrays.asList("application/json"),
                     "User-Agent",
-                    Arrays.asList("aws-sdk-java/2.20.18", "Linux/5.4.0-126")))
+                    Arrays.asList("aws-sdk-java/2.20.131", "Linux/5.4.0-126")))
```
Comment (nastra) [inline aws/src/test/java/org/apache/iceberg/aws/s3/signer/TestS3SignResponseParser.java:77]: no need to update this here

File: python/dev/Dockerfile (hunk 5 above)
Comment (aajisaka) [reply python/dev/Dockerfile]: Reverted. Thanks

File: aws/src/test/java/org/apache/iceberg/aws/s3/signer/TestS3SignResponseParser.java (hunk 6 above)
Comment (aajisaka) [reply aws/src/test/java/org/apache/iceberg/aws/s3/signer/TestS3SignResponseParser.java]: Reverted. Thanks

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIOProperties.java [hunk 7]
```diff
@@ -67,6 +67,14 @@ public class S3FileIOProperties implements Serializable {
    */
   public static final String SSE_TYPE_KMS = "kms";
 
+  /**
+   * S3 DSSE-KMS encryption.
+   *
+   * <p>For more details:
+   * https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingDSSEncryption.html
+   */
+  public static final String DSSE_TYPE_KMS = "kms.dsse";
```
Comment (amogh-jahagirdar) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIOProperties.java:76]: The naming `kms.dsse` implies that DSSE will exist as a property namespace under KMS. Are there any other configurations which would actually go under this namespace? If not should we just call it `dsse-kms`?

Comment (aajisaka) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIOProperties.java]: No, there's any other configuration go under this namespace. Changed to `dsse-kms`.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIOProperties.java (hunk 4 above)
Comment (singhpk234) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIOProperties.java:95]: I see what i meant here was the comment describe an if else structure which makes it a bit hard to follow was wondering if we can re-write this comment in such as way that it's easy for a user to understand / read it. 

This is just a nit and not a blocking comment, as you are just adding a disjunstion to an already existing if branch. please feel free to ignore.

**Review by singhpk234 — APPROVED** [review]
LGTM ! Thanks @aajisaka !

**Comment by aajisaka** [comment]
From my side, is there something do to make it forward?

**Comment by aajisaka** [comment]
Rebased for the latest main branch

**Comment by aajisaka** [comment]
I think this patch is ready to merge:

- Built Iceberg and ran a Spark job using Glue 4.0. I confirmed the Spark job successfully created an Iceberg table under the S3 prefix which is configured to reject any PutObject request without DSSE-KMS encryption.
- Ran integration tests and TestS3FileIOIntegration passed.

cc: @jackye1995 @amogh-jahagirdar

**Comment by jackye1995** [comment]
Was waiting for @nastra , but I agree this is ready to be merged, I will go ahead to do that, thanks for the contribution!

**Comment by aajisaka** [comment]
Thanks!
